File: 0002.add_two_numbers/solution.go
package leetcode0002

import (
	"github.com/Ackerr/Algorithms/utils"
)

func addTwoNumbers(l1 *utils.ListNode, l2 *utils.ListNode) *utils.ListNode {
	ans := &utils.ListNode{Val: -1}
	cur := ans
	flag := 0
	for l1 != nil || l2 != nil || flag == 1 {
		val := flag
		if l1 != nil {
			val += l1.Val
		}
		if l2 != nil {
			val += l2.Val
		}
		flag = val / 10
		cur.Next = &utils.ListNode{Val: val % 10}
		cur = cur.Next
		if l1 != nil {
			l1 = l1.Next
		}
		if l2 != nil {
			l2 = l2.Next
		}
	}
	return ans.Next
}

File: 0002.add_two_numbers/solution.py
from utils import ListNode


class Solution:
    def add_two_numbers(self, l1: ListNode, l2: ListNode) -> ListNode:
        res = ans = ListNode(0)
        pon = 0

        while pon or l1 or l2:
            i = (l1.val if l1 else 0) + (l2.val if l2 else 0) + pon
            pon = i // 10
            ans.next = ListNode(i % 10)
            ans = ans.next
            l1 = l1.next if l1 else None
            l2 = l2.next if l2 else None

        return res.next
